### PR TITLE
DS-3911 Endpoint to delete bitstreams

### PR DESCRIPTION
**Pull request by ppmdo**
Implementation of DELETE on the bitstream objects. The endpoint throws and exception when trying to delete bitstreams whose parent is either a Community or a Collection (logos).

 Has a workaround to avoid blocking of execution by the authorisation manager.

### CODE REVIEW

**Comment by abollini** [comment]
Hi Pablo,
thanks for the PR. The code looks good. Not sure that a 422 is really thrown when attempt to delete a logo as a RTE is thrown I suspect a 500. Can I ask you to add integration tests?
you should at least check that a delete of an existent bitstream will prevent subsequent retrieval 404 and deletetion of a community/collection logo fail with a 422

**Comment by ppmdo** [comment]
Sure. I'll work with the tests. I'm not sure I understood correctly. A deleted bitstream should throw a 404 error when retrieved after deletion?

**Comment by abollini** [comment]
yes.. If the delete succeed the bitstream should be not available anymore

**Comment by ppmdo** [comment]
Hi @abollini. Integration Tests are ready. I've also implemented logic to throw a 404 error when trying to get a deleted bitstream.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java [hunk 1]
```diff
@@ -424,4 +425,82 @@ public void findOneWrongUUID() throws Exception {
...
+            bitstream = BitstreamBuilder.
+                                            createBitstream(context, publicItem1, is)
+                                        .withName("Bitstream")
+                                        .withDescription("Description")
+                                        .withMimeType("text/plain")
+                                        .build();
+        }
+
+        // Verify creation
+        getClient().perform(get("/api/core/bitstreams/" + bitstream.getID()))
```
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java]: This is not necessary as the creation is done via API and not via the REST API

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java [hunk 2]
```diff
@@ -424,4 +425,82 @@ public void findOneWrongUUID() throws Exception {
...
+                    .andExpect(status().is2xxSuccessful());
+
+        // Delete
+        getClient().perform(delete("/api/core/bitstreams/" + bitstream.getID()))
+                .andExpect(status().is2xxSuccessful());
```
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java]: this should be more specific a 204 is expected here

**Review by abollini — CHANGES_REQUESTED** [review]

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamRestRepository.java [hunk 3]
```diff
@@ -89,6 +98,28 @@ public BitstreamResource wrapResource(BitstreamRest bs, String... rels) {
         return new BitstreamResource(bs, utils, rels);
     }
 
+    @Override
+    protected void delete(Context context, UUID id) {
+        // Workaround for authorisation problems
+        context.turnOffAuthorisationSystem();
```
[the remark this answers is not in the document]
Comment (ppmdo) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamRestRepository.java]: Ok. I'll leave this for now.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java (hunk 1 above)
Comment (ppmdo) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java]: Ok. Will remove.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java (hunk 2 above)
Comment (ppmdo) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java]: Ok. Will adjust.

**Comment by ppmdo** [comment]
Hi @abollini. Changes applied.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java [hunk 4]
```diff
@@ -424,4 +425,82 @@ public void findOneWrongUUID() throws Exception {
...
+        // ** GIVEN **
+        // 1. A community with a logo
+        parentCommunity = CommunityBuilder.createCommunity(context).withName("Community").withLogo("logo_community")
+                                          .build();
+
+        // 2. A collection with a logo
+        Collection col = CollectionBuilder.createCollection(context, parentCommunity).withName("Collection")
+                                          .withLogo("logo_collection").build();
+        // Verify parentCommunity logo retreival
+        getClient().perform(get("/api/core/bitstreams/" + parentCommunity.getLogo().getID()))
```
Comment (ppmdo) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java]: Is this unnecessary too?

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java [hunk 5]
```diff
@@ -424,4 +425,82 @@ public void findOneWrongUUID() throws Exception {
...
+                   .andExpect(status().isOk());
+
+        // 422 error when trying to DELETE parentCommunity logo
+        getClient().perform(delete("/api/core/bitstreams/" + parentCommunity.getLogo().getID()))
+                   .andExpect(status().is(422));
+
+        // Verify collection logo retreival
+        getClient().perform(get("/api/core/bitstreams/" + col.getLogo().getID()))
```
Comment (ppmdo) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java]: Same as above. Is this unnecessary?

**Comment by abollini** [comment]
Hi Pablo,
I pushed an additional commit as after further inspection I noted that a better exception handling was required. In your previous code any error will be wrapped in a RuntimeException and reported as a 422. Unfortunately, some broad changes were needed to allow that. @tdonohue can you take a look to these further changes and merge if they make sense for you as well?
I have added additional test to verify proper exception handling of 401 and 403

BTW, I have tested the endpoint with postman and the authorization ignore checkout was not really needed. The issue was indeed related to the HAL browser that doesn't send the Authorization header for methods other than GET, see https://jira.spring.io/browse/DATAREST-1077 we should able to solve that updating our spring data rest dependency

**Review by abollini — APPROVED** [review]

**Comment by ppmdo** [comment]
Indeed, those where broad changes. Thanks for your review! Definitely better exception handling now. Also, I see you're using a token for the auth in the IT tests. I'll take that into account for future reference. :)

**Comment by tdonohue** [comment]
@abollini : This PR looks to currently include unrelated commits that are already on `master`?  I think it may need a rebase before merging, as I don't think it really includes 48 commits.

**Comment by abollini** [comment]
github is confused by my previous merge of the master in this PR (to test it). The commit are the same so once merged they will happen only one time in the history. I prefer to avoid a rebase as it show many conflicts risking to introduce bugs / put large effort for a minimal benefit in a more linear cleaner history

**Pull request merged**